torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -706,31 +701,44 @@ def moe_forward(self, x, topk_ids, topk_weight):
         )
         return final_out
 
-    def moe_on_device(self, x, topk_ids, topk_weight):
+    def sort_tokens(self, x, topk_ids, topk_weights):
         # This part sorts the token indices so that tokens routed to the same expert reside consecutively.
         # An implication is that tokens to the same "expert group" (i.e., device) are also consecutive.
         # Since this is an "aritificial" index creation (final outcome being
         # `idxs`), we don't need gradients here.
         with torch.no_grad():
-            # [seq_len, n_routed_experts]
-            cnts = topk_ids.new_zeros((topk_ids.shape[0], self.config.n_routed_experts))
-            # Fill 1 to the selected experts
-            cnts.scatter_(1, topk_ids, 1)
-            tokens_per_expert = cnts.sum(dim=0)
-            # Token indices for each expert
-            idxs = topk_ids.view(-1).argsort()
-            sorted_tokens_shape = idxs.shape + x.shape[1:]
-
-        sorted_tokens = x[idxs // topk_ids.shape[1]]
-        assert sorted_tokens.shape == sorted_tokens_shape
+            expert_counts = topk_ids.new_zeros(
+                (topk_ids.shape[0], self.config.n_routed_experts)
+            )
+            expert_counts.scatter_(1, topk_ids, 1)
+            tokens_per_expert = expert_counts.sum(dim=0)
+
+            token_indices = topk_ids.view(-1).argsort()
+
+            sorted_tokens_shape = token_indices.shape + x.shape[1:]
 
+        sorted_tokens = x[token_indices // topk_ids.shape[1]]
+        # assert sorted_tokens.shape == sorted_tokens_shape
```

Comment:
no need to return `sorted_tokens_shape`?

Reply:
actually it is used way at the end of the function to filter out the tokens_return_buf.
But, we can just use sorted_tokens.shape[0] and avoid returning it.

Reply:
removed.